### DISCUSSION TITLE
Method to control opacity of any elements ???

### DISCUSSION THREAD

**Discussion by Cicorione** [speedata/publisher#544] [Q&A] [answered]
Hi @pgundlach 

I already shown this code here: https://github.com/speedata/publisher/issues/542#issuecomment-1842128559

Actually that code was made to discuss here… :sweat_smile: 

I haven't found in the documentation a method to control the opacity of any object in any given moment. However, as GUI-guy :man_facepalming:, opacity is a parameter that I expect to have control anytime down the pipe line. Actually with `Publisher` I expect to control it dynamically... :sweat_smile: 

Nevertheless working with the opensource taught me to adapt to the circumstances and I developed a workaround:

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
    xmlns:sd="urn:speedata:2009/publisher/functions/en">

    <DefineColor name="testA" model="RGB" r="255" g="178" b="32" alpha="100" />
    <DefineColor name="testAA" model="RGB" r="255" g="178" b="32" alpha="80" />
    <DefineColor name="testAAA" model="RGB" r="255" g="178" b="32" alpha="60" />
    <DefineColor name="testAAAA" model="RGB" r="255" g="178" b="32" alpha="40" />
 
    <!-- SPColor -->
    <DefineGraphic name="s"> fill box withcolor spcolor(special); </DefineGraphic>

    <Record element="data">
        <SetVariable variable="op" select="'A'" />
        <SetVariable variable="special" select="concat('test',$op)" type="mp:string"/>

        <Loop select="4">
            <PlaceObject>
                <Box height="4" width="4" graphic="s" />
            </PlaceObject>
            <NextRow rows="2"/>
            <SetVariable variable="op" select="concat($op,'A')" />
            <SetVariable variable="special" select="concat('test',$op)" type="mp:string"/>
            <Message select="$op" />
            <Message select="$special" />
        </Loop>
    </Record>
</Layout>
```

I believe this code can be further optimized but it is not the point.

By any chance there is a method to control the opacity that is lesser cumbersome than the one here proposed?

Thanks! :pray:

**Comment by pgundlach** [comment]
I see what you need. But there is currently no way to pass parameters to the metapost instance. I will think of it.

~BTW: your code generates testAA, testAAA, testAAAA and testAAAAA~

edit: your code is correct, except for the message at the end, there is no variable `$special` outside metapost.

**Comment by pgundlach** [comment]
I assume that you want something like this:

```xml
      <Box height="4" width="4" graphic="s">
        <With-Param name="special" select="'testA'" as="mp:string" />
      </Box>
```

(I am not sure about the syntax, this should work with Td for example as well)

**Comment by Cicorione** [comment]
Perhaps the code is not the best but It worked.

In my case I have often highlights info, no more than three by the way, that code will change color to each highlight box creating a nice effect.

**Comment by Cicorione** [comment]
Hi @pgundlach 

I resolved my issue with the "metapost colors" but it looks like my idea to switch color dynamically is not working out as I expected.

The concept is this:
[opacity2.pdf](https://github.com/speedata/publisher/files/13708197/opacity2.pdf)

I hope you are able to spot out the `pebcak` cause right now I haven't understood where it lies.

As on the open threads I have my color variations:
```xml
<DefineColor name="npyellow" model="RGB" r="255" g="178" b="25" alpha="100"/>
<DefineColor name="npyellowA" model="RGB" r="255" g="178" b="25" alpha="80"/>
<DefineColor name="npyellowAA" model="RGB" r="255" g="178" b="25" alpha="60"/>
<DefineColor name="npyellowAAA" model="RGB" r="255" g="178" b="25" alpha="40" />
<DefineColor name="npyellowAAAA" model="RGB" r="255" g="178" b="25" alpha="20" />
```

The variables that control the `accent` attribute are loaded outside the group where the following snippet belongs to:
```xml
<SetVariable variable="accentcolor" select="@accent" />
<SetVariable variable="accent" select="$accentcolor" type="mp:string" />
```

And this is the snippet that should change the color through a `ForAll` loop:
```xml 
<!-- The variable that control accent color -->
<SetVariable variable="extracolor" select="$accentcolor"/>
							
<ForAll select="extra">

	<!-- The variable that control opacity -->
	<SetVariable variable="op" select=""/>
								
	<Message select="concat('Start cycle: ',$extracolor)"/>

	<PlaceObject column="1.3">
		<Frame backgroundcolor="{$extracolor}"
		       framecolor="{$extracolor}"
		       border-bottom-left-radius="2mm"
		       border-bottom-right-radius="2mm"
		       border-top-left-radius="2mm"
		       border-top-right-radius="2mm">
			<Table width="{$noc * 0.7}" padding="2mm">
				<Tr>
					<Td valign="middle" align="center">
						<Paragraph fontfamily="pagenum">
							<Value select="."/>
						</Paragraph>
					</Td>
				</Tr>
			</Table>
		</Frame>
	</PlaceObject>

	<Message select="concat('Mid cycle: ',$extracolor)"/>
	<SetVariable variable="op" select="concat($op,'A')"/>
	<Message select="concat('OP Value: ', $op"/>
	<SetVariable variable="extracolor" select="concat($extracolor,$op)"/>
	<Message select="concat('end cycle: ',$extracolor)"/>

</ForAll>
```

However instead to get same effect as the PDF above I get the same color for each frame:

![image](https://github.com/speedata/publisher/assets/19494201/5cce4c61-a434-4f90-a0a2-d52cfae460d5)

If I look at the `protocol log` of that block, it looks that my dynamic variable is well formed, or at least this is what I believe:

`protocol log`
```sh
PlaceObject: Textblock at (1,19) wd/ht: 8/3 in "unit" (p. 7)
Call "NextRow" from layout
Call "SetVariable" from layout
Call "ForAll" from layout
Call "SetVariable" from layout
Call "Message" from layout
Message: "Start cycle: npyellow"
Call "PlaceObject" from layout
Call "Frame" from layout
Call "Table" from layout
Call "Tr" from layout
Call "Td" from layout
Call "Paragraph" from layout
Call "Value" from layout
PlaceObject: Table at (1,23) wd/ht: 8/4 in "unit" (p. 7)
Call "Message" from layout
Message: "Mid cycle: npyellow"
Call "SetVariable" from layout
Call "Message" from layout
Message: "OP Value: A"
Call "SetVariable" from layout
Call "Message" from layout
Message: "end cycle: npyellowA"
Call "SetVariable" from layout
Call "Message" from layout
Message: "Start cycle: npyellowA"
Call "PlaceObject" from layout
Call "Frame" from layout
Call "Table" from layout
Call "Tr" from layout
Call "Td" from layout
Call "Paragraph" from layout
Call "Value" from layout
PlaceObject: Table at (1,27) wd/ht: 8/4 in "unit" (p. 7)
Call "Message" from layout
Message: "Mid cycle: npyellowA"
Call "SetVariable" from layout
Call "Message" from layout
Message: "OP Value: A"
Call "SetVariable" from layout
Call "Message" from layout
Message: "end cycle: npyellowAA"
Call "SetVariable" from layout
Call "Message" from layout
Message: "Start cycle: npyellowAA"
Call "PlaceObject" from layout
Call "Frame" from layout
Call "Table" from layout
Call "Tr" from layout
Call "Td" from layout
Call "Paragraph" from layout
Call "Value" from layout
PlaceObject: Table at (1,31) wd/ht: 8/4 in "unit" (p. 7)
Call "Message" from layout
Message: "Mid cycle: npyellowAA"
Call "SetVariable" from layout
Call "Message" from layout
Message: "OP Value: A"
Call "SetVariable" from layout
Call "Message" from layout
Message: "end cycle: npyellowAAA"
Call "PlaceObject" from layout
PlaceObject: Group (unit) at (2,1) wd/ht: 1/1 in "_page" (p. 7)
```

It looks to me that `npyellow+A...` is read properly although is not applied, or is applied but it is not applied opacity in that context. Nonetheless, when I passed one color only without any **bizarre idea** opacity was applied properly, thanks! 🙏

**Reply by pgundlach** [reply]
Version 4.15.17 supports opacity for Frame.

**Reply by Cicorione** [reply]
This is pretty cool, but is there any way to control the opacity dynamically? 

Regarding my current issue is the loop malformed or actually the opacity is not read? 

Thanks... 🙏

**Reply by pgundlach** [reply]
Your example works now (4.15.17). Frame colors did not honor opacity until now. There is still no dynamic opacity (I still need to figure out a good interface).

**Reply by Cicorione** [reply]
Ok, let's download the newer version then!

Thank you Patrick, I really like Publisher and you're doing a great job figuring out to make it better while preserving retro-compatibility!

**Reply by Cicorione** [reply]
Ok, it worked but only for `backgroundcolor="{$extracolor}"`, while for `framecolor="{$extracolor}"` not yet... 🤷

![image](https://github.com/speedata/publisher/assets/19494201/aee8ddad-786a-4502-bd88-0888e23922e3)

**Comment by pgundlach** [comment, marked as answer]
Version 4.15.18 will be online in a few minutes and fix this error. Sorry... should have seen that.

**Reply by Cicorione** [reply]
Absolutely not, I am glad that you are following up with my craziness! 🍺

**Reply by Cicorione** [reply]
@pgundlach I'd like to confirm that `4.15.18` fixed the behavior.

I have an additional comment, I noticed that `frame` requires mandatory the border color even if you select a background color, otherwise you get a black border. This is a little bit counter-intuitive and perhaps should be specified in the documentation. Thanks! 🙏
